FLT/ForMathlib/MeasurableSpacePadics.lean

```diff
@@ -0,0 +1,50 @@
+import Mathlib.MeasureTheory.Measure.Haar.Basic
+import Mathlib.NumberTheory.Padics.ProperSpace
+
+open Topology TopologicalSpace MeasureTheory Measure
+
+variable {p : ℕ} [Fact p.Prime]
+
+instance : MeasurableSpace ℚ_[p] := borel _
+instance : BorelSpace ℚ_[p] := ⟨rfl⟩
+
+instance : MeasurableSpace ℤ_[p] := Subtype.instMeasurableSpace
+instance : BorelSpace ℤ_[p] := Subtype.borelSpace _
+
+lemma PadicInt.isOpenEmbedding_coe : IsOpenEmbedding ((↑) : ℤ_[p] → ℚ_[p]) := by
+  refine IsOpen.isOpenEmbedding_subtypeVal (?_ : IsOpen {y : ℚ_[p] | ‖y‖ ≤ 1})
+  simpa only [Metric.closedBall, dist_eq_norm_sub, sub_zero] using
+    IsUltrametricDist.isOpen_closedBall (0 : ℚ_[p]) one_ne_zero
+
+lemma PadicInt.isMeasurableEmbedding_coe : MeasurableEmbedding ((↑) : ℤ_[p] → ℚ_[p]) := by
+  convert PadicInt.isOpenEmbedding_coe.measurableEmbedding
+  exact inferInstanceAs (BorelSpace ℤ_[p])
+
+noncomputable instance : MeasureSpace ℤ_[p] := ⟨addHaarMeasure ⊤⟩
+
+instance : IsAddHaarMeasure (volume : Measure ℤ_[p]) := isAddHaarMeasure_addHaarMeasure _
+
+lemma PadicInt.volume_univ : volume (Set.univ : Set ℤ_[p]) = 1 := addHaarMeasure_self
+
+-- should we more generally make a map from `CompactOpens` to `PositiveCompacts`?
+private def Padic.unitBall_positiveCompact : PositiveCompacts ℚ_[p] where
+  carrier := {y | ‖y‖ ≤ 1}
+  isCompact' := by simpa only [Metric.closedBall, dist_zero_right] using
+    isCompact_closedBall (0 : ℚ_[p]) 1
+  interior_nonempty' := by
+    rw [IsOpen.interior_eq]
+    · exact ⟨0, by simp⟩
+    · simpa only [Metric.closedBall, dist_zero_right] using
+        IsUltrametricDist.isOpen_closedBall (0 : ℚ_[p]) one_ne_zero
+
+noncomputable instance : MeasureSpace ℚ_[p] := ⟨addHaarMeasure Padic.unitBall_positiveCompact⟩
+
+instance : IsAddHaarMeasure (volume : Measure ℚ_[p]) := isAddHaarMeasure_addHaarMeasure _
+
+lemma Padic.volume_closedBall : volume {x : ℚ_[p] | ‖x‖ ≤ 1} = 1 := addHaarMeasure_self
+
+lemma PadicInt.volume_addSubgroupClosure (x : ℤ_[p]ˣ) :
+    volume (AddSubgroup.closure {x.1} : Set ℤ_[p]) = ‖x.1‖₊⁻¹ := sorry
```

Comment:
I think this sorry is false. Firstly, if x is a unit in Z_p then multiplication by x is a bijection Z_p -> Z_p so whatever kind of norm you are putting on x (the p-adic norm, or the one coming from DistribHaarChar etc), the answer will be 1. Conversely, `AddSubgroup.closure {x.1}` will be a countable set so whatever volume you are putting on it, the answer will be 0.

Reply:
Ah I think the issue is that we thought `a • top = closure {a}`, but you are right that the second one only contains very specific sequences of digits, namely the multiples of `a`, while the first one contains many more things.

We can just use `a • top` everywhere

Reply:
This will solve the mathematical problem, but the other thing I'm saying is that `a • top = top` because if u is a unit in a ring R then uR=R because if vu=1 then R = vuR subseteq uR subseteq R

Reply:
Yep, this is my fault. I wanted `u \in nonZeroDivisors \Z_[p]`